server/mscalendar/daily_summary.go

```diff
@@ -120,86 +131,94 @@ func (m *mscalendar) SetDailySummaryEnabled(user *User, enable bool) (*store.Dai
-		}
+	if len(dsumIndex) == 0 {
+		return nil
 	}
 
-	err = m.Store.SaveDailySummaryIndex(dsumIndex)
+	err = m.Filter(withSuperuserClient)
 	if err != nil {
 		return err
 	}
 
-	return nil
-}
+	requests := []*remote.ViewCalendarParams{}
+	for _, dsum := range dsumIndex {
+		shouldPost, err := shouldPostDailySummary(dsum, now)
+		if err != nil {
+			m.Logger.Errorf("Error posting daily summary for user %s: %s", dsum.MattermostUserID, err.Error())
+			continue
+		}
+		if !shouldPost {
+			continue
+		}
 
-func (m *mscalendar) GetDailySummary(user *User) (string, error) {
-	return m.getDailySummary(user)
-}
+		start, end := getTodayHoursForTimezone(now, dsum.Timezone)
+		req := &remote.ViewCalendarParams{
+			RemoteID:  dsum.RemoteID,
+			StartTime: start,
+			EndTime:   end,
+		}
+		requests = append(requests, req)
+	}
 
-func (m *mscalendar) processDailySummary(dsum *store.DailySummarySettings) error {
-	shouldPost, err := shouldPostDailySummary(dsum)
+	responses, err := m.client.DoBatchViewCalendarRequests(requests)
 	if err != nil {
 		return err
 	}
-	if !shouldPost {
-		return nil
-	}
 
-	user := NewUser(dsum.MattermostUserID)
-	postStr, err := m.getDailySummary(user)
-	if err != nil {
-		return err
+	mappedPostTimes := map[string]string{}
+	byRemoteID := dsumIndex.ByRemoteID()
+	for _, res := range responses {
+		dsum := byRemoteID[res.RemoteID]
+		if res.Error != nil {
+			m.Logger.Errorf("Error rendering user %s calendar: %s %s", dsum.MattermostUserID, res.Error.Code, res.Error.Message)
+		}
+		postStr, err := views.RenderCalendarView(res.Events, dsum.Timezone)
+		if err != nil {
+			m.Logger.Errorf("Error rendering user %s calendar: %s", dsum.MattermostUserID, err.Error())
+		}
+
+		m.Poster.DM(dsum.MattermostUserID, postStr)
+		mappedPostTimes[dsum.MattermostUserID] = time.Now().Format(time.RFC3339)
 	}
-	m.Poster.DM(user.MattermostUserID, postStr)
+	m.Logger.Infof("Processed daily summary for %d users", len(mappedPostTimes))
 
-	dsum.LastPostTime = timeNowFunc().Format(time.RFC3339)
+	// TODO atomic save index
 	return nil
 }
 
-func (m *mscalendar) getDailySummary(user *User) (string, error) {
-	tz, err := m.GetTimezone(user)
-	if err != nil {
-		return "", err
-	}
+func (m *mscalendar) GetDailySummaryForUser(user *User) (string, error) {
+	return m.getDailySummaryForUser(user)
+}
 
-	calendarData, err := m.getTodayCalendar(user, tz)
+func (m *mscalendar) getDailySummaryForUser(user *User) (string, error) {
+	tz, err := m.GetTimezone(user)
 	if err != nil {
-		m.Poster.DM(user.MattermostUserID, "Failed to run daily summary job. %s", err.Error())
 		return "", err
 	}
 
-	if len(calendarData) == 0 {
-		m.Poster.DM(user.MattermostUserID, "You have no upcoming events today.")
-		return "", nil
-	}
-
-	postStr, err := views.RenderCalendarView(calendarData, tz)
+	calendarData, err := m.getTodayCalendarEvents(user, time.Now(), tz)
 	if err != nil {
+		m.Poster.DM(user.MattermostUserID, "Failed to get user . %s", err.Error())
```

Comment:
Should this function do the error posting to the user, or should that be handled outside?

Reply:
This is very strange, that's not even the right error message. How about:

```go
return "Failed to get calendar events", err
```

The string return value gets sent back to the user for the slash command response.